calico/openstack/test/test_plugin.py

```diff
@@ -0,0 +1,458 @@
+                #* Advance to the determined time.                           *#
+                #*************************************************************#
+                secs -= (wake_up_time - self.current_time)
+                self.current_time = wake_up_time
+
+                #*************************************************************#
+                #* Wake up all sleepers that should now wake up.             *#
+                #*************************************************************#
+                for queue in self.sleepers.keys():
+                    if self.sleepers[queue] >= self.current_time:
+                        print "Wake up one sleeper: %s" % queue.stack
+                        del self.sleepers[queue]
+                        queue.put_nowait('Wake up!')
+
+                #*************************************************************#
+                #* Allow woken (and possibly other) threads to run.          *#
+                #*************************************************************#
+                self.real_eventlet_sleep(REAL_EVENTLET_SLEEP_TIME)
+
+        #*********************************************************************#
+        #* Hook logging.                                                     *#
+        #*********************************************************************#
+        mech_calico.LOG = mock.Mock()
+        mech_calico.LOG.info.side_effect = log_info
+        mech_calico.LOG.debug.side_effect = log_debug
+        mech_calico.LOG.warn.side_effect = log_warn
+        mech_calico.LOG.exception.side_effect = log_exception
+
+        #*********************************************************************#
+        #* Create an instance of CalicoMechanismDriver.                      *#
+        #*********************************************************************#
+        self.driver = mech_calico.CalicoMechanismDriver()
+
+        #*********************************************************************#
+        #* Hook socket creation and binding.                                 *#
+        #*********************************************************************#
+        mech_calico.zmq.Context = mock.Mock()
+        self.zmq_context = mech_calico.zmq.Context.return_value
+        self.zmq_context.socket.side_effect = socket_created
+
+        #*********************************************************************#
+        #* Hook sleeping.                                                    *#
+        #*********************************************************************#
+        self.real_eventlet_sleep = eventlet.sleep
+        mech_calico.eventlet.sleep = simulated_time_sleep
+
+        #*********************************************************************#
+        #* Tell the driver to initialize.                                    *#
+        #*********************************************************************#
+        self.driver.initialize()
+
```

Comment:
Everything above here is logically setup code, and will likely get duplicated *a lot* for other tests. You should consider pulling it into `TestPlugin.setUp()`.